cpp/src/groupby/sort/group_dense_rank_scan.cu

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/logical.h>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+namespace {
+template <bool nested_nulls>
+void generate_dense_rank_struct_comparisons(column_view const& order_by,
+                                            mutable_column_view out,
+                                            device_span<size_type const> group_labels,
+                                            device_span<size_type const> group_offsets,
+                                            rmm::cuda_stream_view stream)
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<size_type>(),
+    out.end<size_type>(),
+    [d_order_by = *d_order_by,
+     has_nulls  = order_by.has_nulls(),
+     labels     = group_labels.data(),
+     offsets    = group_offsets.data()] __device__(size_type row_index) {
+      if (row_index == offsets[labels[row_index]]) {
+        return true;
+      } else if (has_nulls) {
+        bool const lhs_is_null{d_order_by.is_null_nocheck(row_index)};
+        bool const rhs_is_null{d_order_by.is_null_nocheck(row_index - 1)};
+        if (lhs_is_null and rhs_is_null) {
+          return false;
+        } else if (lhs_is_null != rhs_is_null) {
+          return true;
+        }
+      }
+
+      return !thrust::all_of(
+        thrust::seq,
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(d_order_by.num_child_columns()),
+        [row_index, d_order_by] __device__(size_type child_index) {
+          auto const& col = d_order_by.child(child_index);
+          element_equality_comparator<nested_nulls> element_comparator{col, col, true};
+          return type_dispatcher(col.type(), element_comparator, row_index, row_index - 1);
```

Comment:
Can you try this, please?
```suggestion
          return element_comparator(row_index, row_index - 1);
```

Reply:
If this works, please change it in all other places. I think we should not call type_dispatcher explicitly here.

Reply:
Sorry, shouldn't have resolved the last one, it does not work.

Reply:
Then should use `row_equality_comparator` instead. Here is an example: https://github.com/rapidsai/cudf/blob/7721819eeed68115fd4d7033cba016830b0afcd8/cpp/src/sort/rank.cu#L48

Reply:
`row_equality_comparator` itself calls `type_dispatcher` for you.

Reply:
Commenting this here before I commit/push it because it combines a few different issues that have been brought up. Below @jrhemstad suggested simplifying the code by requiring a struct instead of branching off different functionality for the single column. If I'm switching to `row_equality_comparator` anyway, I can leverage the row comparator's need for a `table_view` to further simplify the code. This version should merge the fractured functionality while handling non-nested structs a little more gracefully.

```
template <bool has_struct_nulls, bool has_nested_nulls>
std::unique_ptr<column> generate_dense_ranks(table_view const& flat_order,
                      column_view const& col_order,
                      rmm::cuda_stream_view stream,
                      rmm::mr::device_memory_resource* mr) {


  auto const d_flat_order = table_device_view::create(flat_order, stream);
  row_equality_comparator<has_nested_nulls> comparator(*d_flat_order, *d_flat_order, true);
  auto ranks = make_fixed_width_column(data_type{type_to_id<size_type>()}, col_order.size(), mask_state::ALL_VALID, stream, mr);
  auto mutable_ranks = ranks->mutable_view();

  if (has_struct_nulls) {
    auto const d_col_order = column_device_view::create(col_order, stream);
    thrust::tabulate(
      rmm::exec_policy(stream),
      mutable_ranks.begin<size_type>(),
      mutable_ranks.end<size_type>(),
        [comparator, d_col_order = *d_col_order] __device__(size_type row_index) {
          if (row_index == 0) {
            return true;
          }
          bool const lhs_is_null{d_col_order.is_null(row_index)};
          bool const rhs_is_null{d_col_order.is_null(row_index-1)};
          if (lhs_is_null and rhs_is_null) {
            return true;
          } else if (lhs_is_null != rhs_is_null) {
            return false;
          }
          return !comparator(row_index, row_index - 1);
        });
  } else {
    thrust::tabulate(
      rmm::exec_policy(stream),
      mutable_ranks.begin<size_type>(),
      mutable_ranks.end<size_type>(),
        [comparator] __device__(
          size_type row_index) {
          return row_index == 0 || !comparator(row_index, row_index - 1);
        });
  }

  thrust::inclusive_scan(rmm::exec_policy(stream),
                         mutable_ranks.begin<size_type>(),
                         mutable_ranks.end<size_type>(),
                         mutable_ranks.begin<size_type>());
  return ranks;
}

std::unique_ptr<column> inclusive_dense_rank_scan(column_view const& order_by,
                                                  rmm::cuda_stream_view stream,
                                                  rmm::mr::device_memory_resource* mr)
{
  if (order_by.type().id() == type_id::STRUCT) {
    table_view flattened{std::vector<column_view>{order_by.child_begin(), order_by.child_end()}};
    if (has_nested_nulls(flattened)) {
      return generate_dense_ranks<true, true>(flattened, order_by, stream, mr);
    } else {
      return generate_dense_ranks<true, false>(flattened, order_by, stream, mr);
    }
  } else if (order_by.has_nulls()) {
    return generate_dense_ranks<false, true>(table_view{{order_by}}, order_by, stream, mr);
  } else {
    return generate_dense_ranks<false, false>(table_view{{order_by}}, order_by, stream, mr);
  }
}
```